File: docs/palindromes-in-peg/scaffold_live_distances.py
"""Compile only live head differences, sharing finite counter registers."""
from gs_head_liveness import analyze
from scaffold_circuit import (Value, Ref, TRUE, FALSE, disjunction)
from scaffold_circuit_structs import Counter, StackPool


class LiveDistances:
  def __init__(self, circuit, program, names, prefix="distance", **analysis_options):
    self.circuit, self.program = circuit, program
    self.analysis = analyze(program, names, **analysis_options)
    self.names = self.analysis.names
    self.pool = StackPool(circuit, {"cells": self.analysis.registers + 1})
    self.registers = [Counter(self.pool, f"{prefix}.r{i}", "cells")
                      for i in range(self.analysis.registers)]
    self.counters = {pair: self.registers[color] for pair, color in self.analysis.colors.items()}
    self.length = Counter(self.pool, prefix + ".load_length", "cells")

  def equal(self, left, right):
    pair, _ = self.analysis.canonical(left, right)
    return TRUE if pair is None else self.counters[pair].zero()

  def less(self, left, right):
    pair, sign = self.analysis.canonical(left, right)
    if pair is None:
      return FALSE
    counter = self.counters[pair]
    return counter.negative() if sign == 1 else counter.positive()

  def load_one(self, enabled):
    self.length.inc(enabled, slot=self.analysis.registers)

  def initialize(self, values, enabled):
    if set(values) != self.analysis.before[self.program.start]:
      raise ValueError("every live entry difference needs an explicit initialization")
    for pair, value in values.items():
      target = self.counters[pair]
      if value is None:
        target.reset(enabled)
      else:
        source, sign = value
        self._assign(target, source.pos, source.neg, sign, enabled)

  @staticmethod
  def _assign(target, positive, negative, sign, enabled):
    if sign == -1:
      positive, negative = negative, positive
    target.pos.copy_from(positive, enabled)
    target.neg.copy_from(negative, enabled)

  def execute(self, events):
    if len(events) != len(self.program.code):
      raise ValueError("one instruction guard per finite state required")
    increments = [[] for _ in self.registers]
    decrements = [[] for _ in self.registers]
    resets, copies = {}, {}
    for state, ((event, _), enabled) in enumerate(zip(self.program.code, events)):
      op = event[0]
      if op == "move":
        if len(event[1]) != 1 or abs(event[1][0][1]) != 1:
          raise ValueError("unit movement required")
        head, amount = event[1][0]
        for pair in self.analysis.after[state]:
          if head in pair:
            change = amount if pair[0] == head else -amount
            (increments if change == 1 else decrements)[self.analysis.colors[pair]].append(enabled)
      elif op == "copy":
        _, target, source = event
        for pair in self.analysis.after[state]:
          if target not in pair:
            continue
          left, right = (source if name == target else name for name in pair)
          original, sign = self.analysis.canonical(left, right)
          destination = self.analysis.colors[pair]
          if original is None:
            resets.setdefault(destination, []).append(enabled)
          else:
            origin = self.analysis.colors[original]
            if destination != origin or sign != 1:
              copies.setdefault((destination, origin, sign), []).append(enabled)
    # Copy sources must be simultaneous pre-instruction values. In particular,
    # two differently named live pairs may exchange their physical registers.
    snapshots = [Counter(self.pool, counter.pos.name.removesuffix(".pos"), "cells")
                 for counter in self.registers]
    for index, counter in enumerate(self.registers):
      counter.inc(disjunction(*increments[index]), slot=index)
      counter.dec(disjunction(*decrements[index]), slot=index)
    for index, conditions in resets.items():
      self.registers[index].reset(disjunction(*conditions))
    for (destination, origin, sign), conditions in copies.items():
      source = snapshots[origin]
      self._assign(self.registers[destination], source.pos, source.neg, sign, disjunction(*conditions))

  def finalize(self):
    for counter in self.registers:
      counter.finalize()
    self.length.finalize()

File: scala/pal/src/main/scala/pal/ScaffoldLiveDistances.scala
package pal

import scala.collection.immutable.VectorMap
import scala.collection.mutable

/** Compile only live head differences, sharing finite counter registers.
  * Port of `scaffold_live_distances.py`; copies use pre-instruction snapshots.
  */
final class LiveDistances(val circuit: Circuit, val program: Program, headNames: Vector[String],
                          prefix: String = "distance", observePositions: Boolean = true,
                          availabilityDistance: Boolean = true) {
  import Expr.{TRUE, FALSE}
  import ScaffoldCircuit.disjunction

  val analysis: Liveness = GsHeadLiveness.analyze(program, headNames, observePositions, availabilityDistance)
  val names: Vector[String] = analysis.names
  val pool = new StackPool(circuit, Vector("cells" -> (analysis.registers + 1)))
  val registers: Vector[Counter] = Vector.tabulate(analysis.registers) { index =>
    new Counter(pool, s"$prefix.r$index", Some("cells"))
  }
  val counters: VectorMap[(String, String), Counter] = VectorMap.from(analysis.colors.iterator.map {
    case (pair, color) => pair -> registers(color)
  })
  val length = new Counter(pool, prefix + ".load_length", Some("cells"))

  def equal(left: String, right: String): Expr = {
    analysis.canonical(left, right)._1.fold(TRUE)(pair => counters(pair).zero())
  }

  def less(left: String, right: String): Expr = {
    val (pair, sign) = analysis.canonical(left, right)
    pair.fold(FALSE) { p => if (sign == 1) counters(p).negative() else counters(p).positive() }
  }

  def loadOne(enabled: Expr): Unit = length.inc(enabled, Some(analysis.registers))

  def initialize(values: collection.Map[(String, String), Option[(Counter, Int)]], enabled: Expr): Unit = {
    if (values.keySet != analysis.before(program.start)) {
      throw new IllegalArgumentException("every live entry difference needs an explicit initialization")
    }
    values.foreach { case (pair, value) =>
      value match {
        case None => counters(pair).reset(enabled)
        case Some((source, sign)) => assign(counters(pair), source.positiveStack, source.negativeStack, sign, enabled)
      }
    }
  }

  private def assign(target: Counter, positive: Stack, negative: Stack, sign: Int, enabled: Expr): Unit = {
    target.positiveStack.copyFrom(if (sign == -1) negative else positive, enabled)
    target.negativeStack.copyFrom(if (sign == -1) positive else negative, enabled)
  }

  def execute(events: Seq[Expr]): Unit = {
    if (events.size != program.code.size) {
      throw new IllegalArgumentException("one instruction guard per finite state required")
    }
    val increments = Vector.fill(registers.size)(mutable.ArrayBuffer.empty[Expr])
    val decrements = Vector.fill(registers.size)(mutable.ArrayBuffer.empty[Expr])
    val resets = mutable.LinkedHashMap.empty[Int, mutable.ArrayBuffer[Expr]]
    val copies = mutable.LinkedHashMap.empty[(Int, Int, Int), mutable.ArrayBuffer[Expr]]
    program.code.zip(events).zipWithIndex.foreach { case ((row, enabled), state) =>
      row.event match {
        case Event.Move(moves) =>
          if (moves.size != 1 || math.abs(moves.head.delta) != 1) {
            throw new IllegalArgumentException("unit movement required")
          }
          val Event.Movement(head, amount) = moves.head
          analysis.after(state).foreach { pair =>
            if (pair._1 == head || pair._2 == head) {
              val change = if (pair._1 == head) amount else -amount
              (if (change == 1) increments else decrements)(analysis.colors(pair)) += enabled
            }
          }
        case Event.Copy(target, source) =>
          analysis.after(state).foreach { pair =>
            if (pair._1 == target || pair._2 == target) {
              val (original, sign) = analysis.canonical(
                if (pair._1 == target) source else pair._1, if (pair._2 == target) source else pair._2)
              val destination = analysis.colors(pair)
              original match {
                case None => resets.getOrElseUpdate(destination, mutable.ArrayBuffer.empty) += enabled
                case Some(from) =>
                  val origin = analysis.colors(from)
                  if (destination != origin || sign != 1) {
                    copies.getOrElseUpdate((destination, origin, sign), mutable.ArrayBuffer.empty) += enabled
                  }
              }
            }
          }
        case _ => ()
      }
    }
    val snapshots = registers.map(counter => new Counter(pool, counter.positiveStack.name.stripSuffix(".pos"), Some("cells")))
    registers.zipWithIndex.foreach { case (counter, index) =>
      counter.inc(disjunction(increments(index).toSeq*), Some(index))
      counter.dec(disjunction(decrements(index).toSeq*), Some(index))
    }
    resets.foreach { case (index, conditions) => registers(index).reset(disjunction(conditions.toSeq*)) }
    copies.foreach { case ((destination, origin, sign), conditions) =>
      val source = snapshots(origin)
      assign(registers(destination), source.positiveStack, source.negativeStack, sign, disjunction(conditions.toSeq*))
    }
  }

  def commit(): Unit = { registers.foreach(_.commit()); length.commit() }
}
